### ISSUE TITLE
Issue with skipping imbalanced split while running MnistMLPExample on spark cluster

### ISSUE THREAD

**Issue by omerwase** [deeplearning4j/deeplearning4j#3134]
## I'm using a standalone spark cluster:
- 3 AWS m4.xlarge instances (1 master+driver, 2 workers)
- All have java 1.8.0_121 and spark-1.6.0-bin-hadoop2.6
- I start master and join workers manually. Here is my spark-env.sh file:
   - SPARK_DAEMON_MEMORY=2G
   - SPARK_WORKER_DIR=/home/ec2-user/spark/temp
   - SPARK_WORKER_CORES=4
   - SPARK_WORKER_MEMORY=12G
   - SPARK_EXECUTOR_MEMORY=6G
   - SPARK_EXECUTOR_CORES=2
   - SPARK_DAEMON_JAVA_OPTS="-D-XX:+UseG1GC"
   - SPARK_LOCAL_DIRS=/home/ec2-user/spark/temp2

## When I run:
`./bin/spark-submit --driver-memory 6G --class org.deeplearning4j.mlp.MnistMLPExample --master spark://..:7077 bin/files/dl4j-spark-0.7-SNAPSHOT-bin.jar -useSparkLocal false`

## I get the following output:

> 05:24:17,854 INFO  ~ Loaded [CpuBackend] backend
05:24:17,901 INFO  ~ Number of threads used for NativeOps: 4
05:24:26,650 WARN  ~ Warning: new network default sets pretrain to false.
05:24:26,650 WARN  ~ Warning: new network default sets backprop to true.
05:24:26,926 INFO  ~ Initiating RDD<DataSet> export at /tmp/hadoop-ec2-user/dl4j/1490678666675_-55f0d93/0/
05:24:44,016 INFO  ~ RDD<DataSet> export complete at /tmp/hadoop-ec2-user/dl4j/1490678666675_-55f0d93/0/
05:24:44,125 INFO  ~ Starting training of split 1 of 1. workerMiniBatchSize=16, averagingFreq=5, Configured for 8 workers
05:24:48,828 INFO  ~ Skipping imbalanced split with no data for all executors   
05:24:48,828 INFO  ~ Completed training of split 1 of 1
05:24:48,829 INFO  ~ Completed Epoch 0
05:24:48,868 INFO  ~ Starting training of split 1 of 1. workerMiniBatchSize=16, averagingFreq=5, Configured for 8 workers
05:24:49,046 INFO  ~ Skipping imbalanced split with no data for all executors
05:24:49,046 INFO  ~ Completed training of split 1 of 1
05:24:49,046 INFO  ~ Completed Epoch 1
05:24:49,083 INFO  ~ Starting training of split 1 of 1. workerMiniBatchSize=16, averagingFreq=5, Configured for 8 workers
05:24:49,184 INFO  ~ Skipping imbalanced split with no data for all executors
05:24:49,184 INFO  ~ Completed training of split 1 of 1
05:24:49,184 INFO  ~ Completed Epoch 2
05:24:49,210 INFO  ~ Starting training of split 1 of 1. workerMiniBatchSize=16, averagingFreq=5, Configured for 8 workers
05:24:49,311 INFO  ~ Skipping imbalanced split with no data for all executors
05:24:49,311 INFO  ~ Completed training of split 1 of 1
05:24:49,311 INFO  ~ Completed Epoch 3
05:24:49,338 INFO  ~ Starting training of split 1 of 1. workerMiniBatchSize=16, averagingFreq=5, Configured for 8 workers
05:24:49,604 INFO  ~ Skipping imbalanced split with no data for all executors
05:24:49,604 INFO  ~ Completed training of split 1 of 1
05:24:49,604 INFO  ~ Completed Epoch 4
05:24:49,633 INFO  ~ Starting training of split 1 of 1. workerMiniBatchSize=16, averagingFreq=5, Configured for 8 workers
05:24:49,742 INFO  ~ Skipping imbalanced split with no data for all executors
05:24:49,742 INFO  ~ Completed training of split 1 of 1
05:24:49,742 INFO  ~ Completed Epoch 5
05:24:49,771 INFO  ~ Starting training of split 1 of 1. workerMiniBatchSize=16, averagingFreq=5, Configured for 8 workers
05:24:49,862 INFO  ~ Skipping imbalanced split with no data for all executors
05:24:49,862 INFO  ~ Completed training of split 1 of 1
05:24:49,862 INFO  ~ Completed Epoch 6
05:24:49,887 INFO  ~ Starting training of split 1 of 1. workerMiniBatchSize=16, averagingFreq=5, Configured for 8 workers
05:24:49,992 INFO  ~ Skipping imbalanced split with no data for all executors
05:24:49,992 INFO  ~ Completed training of split 1 of 1
05:24:49,992 INFO  ~ Completed Epoch 7
05:24:50,030 INFO  ~ Starting training of split 1 of 1. workerMiniBatchSize=16, averagingFreq=5, Configured for 8 workers
05:24:50,237 INFO  ~ Skipping imbalanced split with no data for all executors
05:24:50,237 INFO  ~ Completed training of split 1 of 1
05:24:50,237 INFO  ~ Completed Epoch 8
05:24:50,263 INFO  ~ Starting training of split 1 of 1. workerMiniBatchSize=16, averagingFreq=5, Configured for 8 workers
05:24:50,390 INFO  ~ Skipping imbalanced split with no data for all executors
05:24:50,391 INFO  ~ Completed training of split 1 of 1
05:24:50,391 INFO  ~ Completed Epoch 9
05:24:50,414 INFO  ~ Starting training of split 1 of 1. workerMiniBatchSize=16, averagingFreq=5, Configured for 8 workers
05:24:50,515 INFO  ~ Skipping imbalanced split with no data for all executors
05:24:50,516 INFO  ~ Completed training of split 1 of 1
05:24:50,516 INFO  ~ Completed Epoch 10
05:24:50,541 INFO  ~ Starting training of split 1 of 1. workerMiniBatchSize=16, averagingFreq=5, Configured for 8 workers
05:24:50,660 INFO  ~ Skipping imbalanced split with no data for all executors
05:24:50,660 INFO  ~ Completed training of split 1 of 1
05:24:50,660 INFO  ~ Completed Epoch 11
05:24:50,684 INFO  ~ Starting training of split 1 of 1. workerMiniBatchSize=16, averagingFreq=5, Configured for 8 workers
05:24:50,803 INFO  ~ Skipping imbalanced split with no data for all executors
05:24:50,803 INFO  ~ Completed training of split 1 of 1
05:24:50,803 INFO  ~ Completed Epoch 12
05:24:50,826 INFO  ~ Starting training of split 1 of 1. workerMiniBatchSize=16, averagingFreq=5, Configured for 8 workers
05:24:50,906 INFO  ~ Skipping imbalanced split with no data for all executors
05:24:50,906 INFO  ~ Completed training of split 1 of 1
05:24:50,906 INFO  ~ Completed Epoch 13
05:24:50,930 INFO  ~ Starting training of split 1 of 1. workerMiniBatchSize=16, averagingFreq=5, Configured for 8 workers
05:24:51,009 INFO  ~ Skipping imbalanced split with no data for all executors
05:24:51,009 INFO  ~ Completed training of split 1 of 1
05:24:51,009 INFO  ~ Completed Epoch 14
05:25:49,862 INFO  ~ ***** Evaluation *****                                     
05:25:49,863 INFO  ~ 
Examples labeled as 0 classified by model as 0: 54 times
Examples labeled as 0 classified by model as 1: 139 times
Examples labeled as 0 classified by model as 2: 709 times
Examples labeled as 0 classified by model as 4: 83 times
Examples labeled as 0 classified by model as 6: 29 times
Examples labeled as 0 classified by model as 8: 4909 times
Examples labeled as 1 classified by model as 0: 4 times
Examples labeled as 1 classified by model as 1: 852 times
Examples labeled as 1 classified by model as 2: 13 times
Examples labeled as 1 classified by model as 4: 3 times
Examples labeled as 1 classified by model as 6: 22 times
Examples labeled as 1 classified by model as 7: 1 times
Examples labeled as 1 classified by model as 8: 5847 times
Examples labeled as 2 classified by model as 0: 20 times
Examples labeled as 2 classified by model as 1: 923 times
Examples labeled as 2 classified by model as 2: 319 times
Examples labeled as 2 classified by model as 4: 30 times
Examples labeled as 2 classified by model as 5: 4 times
Examples labeled as 2 classified by model as 6: 7 times
Examples labeled as 2 classified by model as 8: 4655 times
Examples labeled as 3 classified by model as 0: 30 times
Examples labeled as 3 classified by model as 1: 368 times
Examples labeled as 3 classified by model as 2: 148 times
Examples labeled as 3 classified by model as 4: 67 times
Examples labeled as 3 classified by model as 5: 1 times
Examples labeled as 3 classified by model as 6: 13 times
Examples labeled as 3 classified by model as 7: 26 times
Examples labeled as 3 classified by model as 8: 5468 times
Examples labeled as 3 classified by model as 9: 10 times
Examples labeled as 4 classified by model as 0: 14 times
Examples labeled as 4 classified by model as 1: 1039 times
Examples labeled as 4 classified by model as 2: 60 times
Examples labeled as 4 classified by model as 4: 349 times
Examples labeled as 4 classified by model as 6: 74 times
Examples labeled as 4 classified by model as 7: 8 times
Examples labeled as 4 classified by model as 8: 4297 times
Examples labeled as 4 classified by model as 9: 1 times
Examples labeled as 5 classified by model as 0: 174 times
Examples labeled as 5 classified by model as 1: 405 times
Examples labeled as 5 classified by model as 2: 704 times
Examples labeled as 5 classified by model as 3: 1 times
Examples labeled as 5 classified by model as 4: 170 times
Examples labeled as 5 classified by model as 5: 1 times
Examples labeled as 5 classified by model as 6: 59 times
Examples labeled as 5 classified by model as 7: 12 times
Examples labeled as 5 classified by model as 8: 3894 times
Examples labeled as 5 classified by model as 9: 1 times
Examples labeled as 6 classified by model as 0: 23 times
Examples labeled as 6 classified by model as 1: 680 times
Examples labeled as 6 classified by model as 2: 513 times
Examples labeled as 6 classified by model as 4: 63 times
Examples labeled as 6 classified by model as 5: 3 times
Examples labeled as 6 classified by model as 6: 26 times
Examples labeled as 6 classified by model as 8: 4610 times
Examples labeled as 7 classified by model as 0: 104 times
Examples labeled as 7 classified by model as 1: 607 times
Examples labeled as 7 classified by model as 2: 149 times
Examples labeled as 7 classified by model as 4: 120 times
Examples labeled as 7 classified by model as 6: 35 times
Examples labeled as 7 classified by model as 7: 54 times
Examples labeled as 7 classified by model as 8: 5189 times
Examples labeled as 7 classified by model as 9: 7 times
Examples labeled as 8 classified by model as 0: 22 times
Examples labeled as 8 classified by model as 1: 700 times
Examples labeled as 8 classified by model as 2: 96 times
Examples labeled as 8 classified by model as 3: 1 times
Examples labeled as 8 classified by model as 4: 142 times
Examples labeled as 8 classified by model as 5: 1 times
Examples labeled as 8 classified by model as 6: 5 times
Examples labeled as 8 classified by model as 7: 3 times
Examples labeled as 8 classified by model as 8: 4881 times
Examples labeled as 9 classified by model as 0: 51 times
Examples labeled as 9 classified by model as 1: 1042 times
Examples labeled as 9 classified by model as 2: 62 times
Examples labeled as 9 classified by model as 4: 104 times
Examples labeled as 9 classified by model as 6: 29 times
Examples labeled as 9 classified by model as 7: 53 times
Examples labeled as 9 classified by model as 8: 4608 times
==========================Scores========================================
 Accuracy:        0.1089
 Precision:       0.129
 Recall:          0.1096
 F1 Score:        0.1185
========================================================================
05:25:49,863 INFO  ~ Attempting to delete temporary directory: /tmp/hadoop-ec2-user/dl4j/1490678666675_-55f0d93/0/
05:25:49,864 INFO  ~ Deleted temporary directory: /tmp/hadoop-ec2-user/dl4j/1490678666675_-55f0d93/0/
05:25:49,864 INFO  ~ ***** Example Complete *****

## Here is the spark log file from master (see 2 WARN at the end):

> Spark Command: /usr/java/default/bin/java -cp /home/ec2-user/spark/conf/:/home/ec2-user/spark/lib/spark-assembly-1.6.0-hadoop2.6.0.jar:/home/ec2-user/spark/lib/datanucleus-core-3.2.10.jar:/home/ec2-user/spark/lib/datanucleus-rdbms-3.2.9.jar:/home/ec2-user/spark/lib/datanucleus-api-jdo-3.2.6.jar -D-XX:+UseG1GC -Xms6G -Xmx6G org.apache.spark.deploy.master.Master --ip ... --port 7077 --webui-port 8080
========================================
Using Spark's default log4j profile: org/apache/spark/log4j-defaults.properties
17/03/28 05:38:00 INFO Master: Registered signal handlers for [TERM, HUP, INT]
17/03/28 05:38:00 WARN NativeCodeLoader: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable
17/03/28 05:38:00 INFO SecurityManager: Changing view acls to: ec2-user
17/03/28 05:38:00 INFO SecurityManager: Changing modify acls to: ec2-user
17/03/28 05:38:00 INFO SecurityManager: SecurityManager: authentication disabled; ui acls disabled; users with view permissions: Set(ec2-user); users with modify permissions: Set(ec2-user)
17/03/28 05:38:01 INFO Utils: Successfully started service 'sparkMaster' on port 7077.
17/03/28 05:38:01 INFO Master: Starting Spark master at spark://...:7077
17/03/28 05:38:01 INFO Master: Running Spark version 1.6.0
17/03/28 05:38:01 INFO Utils: Successfully started service 'MasterUI' on port 8080.
17/03/28 05:38:01 INFO MasterWebUI: Started MasterWebUI at http://...:8080
17/03/28 05:38:01 INFO Utils: Successfully started service on port 6066.
17/03/28 05:38:01 INFO StandaloneRestServer: Started REST server for submitting applications on port 6066
17/03/28 05:38:01 INFO Master: I have been elected leader! New state: ALIVE
17/03/28 05:38:04 INFO Master: Registering worker ...:42666 with 4 cores, 14.0 GB RAM
17/03/28 05:38:07 INFO Master: Registering worker ...:41569 with 4 cores, 14.0 GB RAM
17/03/28 05:38:51 INFO Master: Registering app DL4J Spark MLP Example
17/03/28 05:38:51 INFO Master: Registered app DL4J Spark MLP Example with ID app-20170328053851-0000
17/03/28 05:38:51 INFO Master: Launching executor app-20170328053851-0000/0 on worker worker-20170328053803-172.31.51.247-42666
17/03/28 05:38:51 INFO Master: Launching executor app-20170328053851-0000/1 on worker worker-20170328053807-172.31.50.68-41569
17/03/28 05:39:39 INFO Master: Received unregister request from application app-20170328053851-0000
17/03/28 05:39:39 INFO Master: Removing app app-20170328053851-0000
17/03/28 05:39:39 INFO Master: ip-172-31-60-81.ec2.internal:38350 got disassociated, removing it.
17/03/28 05:39:39 INFO Master: 172.31.60.81:45460 got disassociated, removing it.
17/03/28 05:39:40 WARN Master: Got status update for unknown executor app-20170328053851-0000/1
17/03/28 05:39:40 WARN Master: Got status update for unknown executor app-20170328053851-0000/0

**Comment by AlexDBlack** [comment]
With an averaging frequency of 5, 8 workers, minibatch size of 16: that gives 640 examples per split, and hence 60k/640 = 94 splits per epoch.
You're getting 1 split per epoch. So something is definitely up with your data - it's like it's being fit with no data.

Can you confirm:
(a) This is running with DL4J version 0.8.0
(b) The example itself is unmodified

Beyond that: I'd have to run this on AWS to try to reproduce it. Though in the past, I've extensively used DL4J on EC2 without any issues.

**Comment by AlexDBlack** [comment]
Oh, and one thing you can check: 
```
05:24:44,016 INFO ~ RDD export complete at /tmp/hadoop-ec2-user/dl4j/1490678666675_-55f0d93/0/
```

If you've still got the instances running (or, happen to try it again) can you post what is actually in that directory?
I'm thinking this directory is not network accessible - i.e., it's a local file system, not a shared file system like HDFS?

**Comment by omerwase** [comment]
Hi Alex, thank you for your help. I will follow your instructions and provide more information.

**Comment by agibsonccc** [comment]
Closing: Spark is no longer supported in DL4J.
